rfc/0011-audit-checks.md

```diff
@@ -0,0 +1,354 @@
+We iterate over each one of them and create a dictionary that has:
+
+* Key: the Kubernetes resource type, for example `Pod`
+* Value: a list of policies that evaluate the given Kubernetes resource
+
+The idea is to have a map of policies that tell us these information:
+
+* `Pod` resources are relevant to `polA`, `polB`
+* `Deployment` resources are relevant to `polA`
+* `Ingress` resources are relevant to `polC`
+
+### Step 4: fetch Kubernetes resources to be audited
+
+Next we iterate over the keys of this dictionary and, for each one of them we
+query the API server to get all the resources of type `X` defined inside of our Namespace.
+
+For example, given the previous example, we would end up with:
+
+* List of all the `Pod` objects defined inside of the audited Namespace
+* List of all the `Deployment` objects defined inside of the audited Namespace
+* List of all the `Ingress` objects defined inside of the audited Namespace
+
+**Note:** there's no need to do any caching of the response we get from the Kubernetes API server because of the following reasons:
+
+* We work on a per Namespace basis, these query results are useful only in the context of the Namespace being audited
+* The audit reports is generated every X minutes (maybe 10/30/60 minutes), cached data would be useless after this time
+
+### Step 5: attempt to reuse previous evaluation results
+
+Now we have the following:
+
+* List of policies interested by a certain type of Kubernetes resource, like `Pod`
+* List of Kubernetes resources, `Pod` in this case, defined inside of the Namespace we are auditing
+
+The code will then retrieve the `PolicyReport` for the namespace being audited.
+
+We now iterate over each resource, for example over each `Pod` resource defined
+inside of the Namespace, and perform the following operations:
+
+* Iterate over all the policies that are interested in `Pod` resources, for
+  each tuple of (policy `X`, pod instance `Y`):
+  * Look into the `PolicyReport` for an entry that features policy `X`
+    evaluating pod instance `Y`:
+    * If there's an entry, look at the evaluation timestamp of this entry.
+      If policy update timestamp **and** pod instance update timestamp are older
+      than the evaluation timestamp it means that nothing changed. Hence we are
+      going to reuse the old evaluation result. Otherwise something changed,
+      hence we **might** need to perform an evaluation.
+    * If there's no entry we **might** perform an evaluation
+  * If we reached this stage, it means the pod instance `Y` might need to be
+    evaluated by policy `X`. We have to check the `spec.ObjectSelector` field
+    defined inside of the policy (this applies both to `ClusterAdmissionPolicy`
+    and to `AdmissionPolicy`) and figure out if the `ObjectSelector` is
+    compatible with the pod instance `Y`. If that's true, we are going to
+    perform the actual evaluation
```

Comment:
@raulcabello this is where the `objectSelector` is taken into account. IF and only IF we can determine whether pod instance `Y` is compatible with the given `objectSelector` **without** doing another API call.

If we have to perform an API call, then we should revisit the two steps and, for example, make multiple API requests to get the exact list of pod instances that are relevant to each policy

Reply:
perfect! I think we can use the [Matches](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L51) method without doing any extra API call